.github/workflows/ci-circt-nightly.yml

```diff
@@ -16,12 +16,102 @@ on:
   # Run on PRs that target the
   pull_request:
     branches:
-      - ${{ branch-name }}
+      - ci/ci-circt-nightly
 
 jobs:
-  hello:
-    name: Hello World
+  update-branch:
+    name: Integrate staging branch with default branch
     runs-on: ubuntu-20.04
     steps:
-      - name: Hello
-        run: echo "World"
+      - name: Checkout
+        uses: actions/checkout@v3
+        with:
+          ref: main
+      - run: |
+          # Leave the staging branch unchanged if this is a PR.
+          if [[ ${{ github.event_name }} == 'pull_request' ]]; then
+            exit 0
+          fi
+
+          # Rebase the staging branch on the default branch.  Create the staging
+          # branch if it doesn't exist.
+          git checkout -b ${{ env.branch-name }}
+          if [[ `git fetch origin ${{ env.branch-name }}` ]]; then
+            git reset --hard origin ${{ env.branch-name }}
+            git rebase origin/main
+          fi
+
+          git push --force origin ${{ env.branch-name }}
+
+  # If this is running on a PR, then use the branch of the PR.  Otherwise, use
+  # the staging branch.
+  determine-branch:
+    name: Determine what branch to run tests on
+    runs-on: ubuntu-20.04
+    steps:
+      - name: Determine Branch
+        id: determine-branch
+        run: |
+          if [[ ${{ github.event_name }} != 'pull_request' ]]; then
+            echo branches='["${{ env.branch-name }}"]' >> $GITHUB_OUTPUT
+          else
+            echo branches='["${{ github.ref }}"]' >> $GITHUB_OUTPUT
+          fi
+    outputs:
+      branches: ${{ steps.determine-branch.outputs.branches }}
+
+  ci:
+    name: ci
+    needs: [update-branch, determine-branch]
+    strategy:
+      matrix:
+        system: ["ubuntu-20.04"]
+        jvm: [8]
+        scala: ["2.13.11"]
+        espresso: ["2.4"]
+        circt: ["nightly"]
+        ref: ${{ fromJSON(needs.determine-branch.outputs.branches) }}
```

Comment:
For my own understanding, are you using JSON here just so that you're able to include ref in the build matrix so that it shows up in the Github UI?

Reply:
That would require this construction, though it's not the immediate problem this is solving.

I need to configure the build matrix. More accurately, I need to tell `test.yml` which branch to use. Either use whatever the current branch is (for the PR flow) or switch to the staging branch for the nightly flow. The previous job (`determine-branch`) figures out what the branch should be. Funneling this through JSON is the only way that I know how to do this.

Relatedly, the `env` context is not available in the `matrix` (I think). So, you wind up with a similar construction if you want to only define the branch name in one place.